test/correctness/atomics.cpp

```diff
@@ -0,0 +1,1223 @@
+        Buffer<int> out = consumer.realize(hist_size);
+        for (int i = 0; i < hist_size; i++) {
+            check(__LINE__, out(i), correct(i));
+        }
+    }
+}
+
+void test_async_tuple(const Backend &backend) {
+    int img_size = 10000;
+    int hist_size = 7;
+
+    Func producer0, producer1, consumer0, consumer1;
+    Var x;
+    RDom r(0, img_size);
+    RDom rh(0, hist_size);
+
+    producer0(x) = (x*x) % hist_size;
+    producer1(x) = ((x+1)*(x-1)) % hist_size;
+
+    consumer0(x) = Tuple(0, 0);
+    consumer0(producer0(r)) += Tuple(1, 1);
+    consumer0(producer1(r)) += Tuple(1, 1);
+
+    consumer1(x) = Tuple(0, 0);
+    consumer1(clamp(consumer0(rh)[0], 0, 2*img_size)) += Tuple(1, 1);
+
+    consumer0.compute_root().async();
+    producer0.compute_root().async().parallel(x);
+    producer1.compute_root().async().parallel(x);
+    consumer1.compute_root();
+    switch(backend) {
+        case Backend::CPU: {
+            consumer0.update(0)
+                     .atomic(true /*override_associativity_test*/)
+                     .parallel(r);
+            consumer0.update(1)
+                     .atomic(true /*override_associativity_test*/)
+                     .parallel(r);
+            consumer1.update()
+                     .atomic()
+                     .parallel(rh);
+        } break;
+        default: {
+            _halide_user_assert(false) << "Unsupported backend.\n";
+        } break;
+    }
+
+    Buffer<int> correct_consumer0(hist_size);
+    Buffer<int> correct_consumer1(2 * img_size);
+    correct_consumer0.fill(0);
+    correct_consumer1.fill(0);
+    for (int i = 0; i < img_size; i++) {
+        int idx = (i*i) % hist_size;
+        correct_consumer0(idx) = correct_consumer0(idx) + 1;
+    }
+    for (int i = 0; i < img_size; i++) {
+        // Halide's modulo behaves differently compared to C's modulo.
+        int idx = Halide::Internal::mod_imp(((i+1)*(i-1)), hist_size);
+        correct_consumer0(idx) = correct_consumer0(idx) + 1;
+    }
+    for (int i = 0; i < hist_size; i++) {
+        int idx = correct_consumer0(i);
+        correct_consumer1(idx) = correct_consumer1(idx) + 1;
+    }
+
+    // Run 10 times to make sure race condition do not happen
+    for (int iter = 0; iter < 10; iter++) {
+        Realization out = consumer1.realize(2 * img_size);
+        Buffer<int> out0 = out[0];
+        Buffer<int> out1 = out[1];
+        for (int i = 0; i < 2 * img_size; i++) {
+            check(__LINE__, out0(i), correct_consumer1(i));
+            check(__LINE__, out1(i), correct_consumer1(i));
+        }
+    }
+}
+
+int main(int argc, char **argv) {
+    Target target = get_jit_target_from_environment();
+    // Most of the schedules used in this test are terrible for large
+    // thread count machines, due to massive amounts of
+    // contention. We'll just set the thread count to 4. Unfortunately
+    // there's no JIT api for this yet.
+    #ifdef _WIN32
+    _putenv_s("HL_NUM_THREADS", "4");
```

Comment:
Does `halide_set_num_threads()` not work for the JIT? If not, that's a major oversight. (Could you open an issue on it?)

Reply:
It looks like it should works. @abadams is it fine if I change it to use `halide_set_num_threads`?

Reply:
I don't think it works for JIT. It's a runtime function. I went looking for an accessor but could not find it. Should probably be a method on JITSharedRuntime

Reply:
OK. Opened an issue.